File: android-obd-library/src/main/java/com/obd/lib/statics/Translations.java
package com.obd.lib.statics;

import com.bradbarnhill.obd.android.R;
import com.obd.lib.models.PID;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.List;

/**
 * This class provides translations for pids that are based on enumeration or other translation
 * methods other than just conversion using a simple formula.
 *
 * @author Brad Barnhill
 */
public class Translations {
    public static boolean HandleSpecialPidEnumerations(final PID pid, final ArrayList<Short> buffer) {
        if (pid.Mode.equals("01") && pid.PID.equals("00")) {
            mode1Pid00_Translation(pid, buffer);
            return true;
        } else if (pid.Mode.equals("01") && pid.PID.equals("01")) {
            //dtc count
            mode1Pid01_Translation(pid, buffer);
            return true;
        } else if (pid.Mode.equals("01") && pid.PID.equals("51")) {
            //fuel type
            mode1Pid51_Translation(pid, buffer);
            return true;
        }

        return false;
    }

    /**
     * Not finished yet
     */
    private static void mode1Pid00_Translation(final PID pid, final ArrayList<Short> buffer) {
        if (buffer.size() != 6) {
            return;
        }

        String hexString = "";
        final List<Short> subList = buffer.subList(2, buffer.size());
        String temp;
        for (int i = 0; i < subList.size(); i++) {
            temp = Integer.toHexString(subList.get(i));
            hexString += (temp.length() == 1 ? "0" : "") + temp;
        }

        pid.CalculatedResult = Long.parseLong(hexString, 16);
        pid.CalculatedResultString = hexString;
    }

    private static void mode1Pid01_Translation(final PID pid, final ArrayList<Short> buffer) {
        if (!buffer.isEmpty() && buffer.size() > 2) {
            // ignore first two bytes [hh hh] of the response
            final int mil = buffer.get(2);
            final String on = AppContext.getResourceString(R.string.mode1_pid01_translation_on);
            final String off = AppContext.getResourceString(R.string.mode1_pid01_translation_off);
            final String onOff = (mil & 0x80) == 128 ? on : off;
            pid.CalculatedResultString = String.format(AppContext.getResourceString(R.string.mode1_pid01_translation), onOff, mil & 0x7F);
        }
    }

    private static void mode1Pid51_Translation(final PID pid, final ArrayList<Short> buffer) {
        final String[] pidTranslation = AppContext.getResourceStringArray(R.array.mode1_pid51_translation);
        if (!buffer.isEmpty() && buffer.get(2) < pidTranslation.length) {
            pid.CalculatedResultString = pidTranslation[buffer.get(2)];
        }
    }

    private final static int bitsInOneCharValue = 4;
    private final static int HEX_RADIX = 16;
    private final static String SPACE = " ";
    /**
     * Convert the given hexadecimal number to bits
     *
     * @param hex The {@link String} representation of the hexadecimal
     * @return {@link BitSet} containing the bits of the hexadecimal
     */
    public static BitSet hexToBitSet(final String hex) {
        final int hexBitSize = hex.length() * bitsInOneCharValue;
        final BitSet hexBitSet = new BitSet(hexBitSize);

        for (int i = 0; i < hex.length(); i++) {
            final Character hexChar = hex.charAt(i);
            final BitSet charBitSet = hexCharToBitSet(hexChar);

            for (int j = 0; j < bitsInOneCharValue; j++) {
                if (charBitSet.get(j)) {
                    hexBitSet.set(j + (hex.length() - i - 1) * bitsInOneCharValue);
                }
            }
        }

        return hexBitSet;
    }

    /**
     * Convert the given hexadecimal character to its bits. Note: valid inputs
     * are 0-F
     *
     * @param hexChar the hexadecimal character to convert
     * @return BitSet containing the bits of the hexadecimal character
     */
    public static BitSet hexCharToBitSet(final Character hexChar) {
        final BitSet charBitSet = new BitSet(bitsInOneCharValue);
        final int hex = Integer.parseInt(hexChar.toString(), HEX_RADIX);

        for (int i = 0; i < bitsInOneCharValue; i++) {
            final int bit = Integer.lowestOneBit(hex >> i);
            if (bit == 1)
                charBitSet.set(i);
        }

        return charBitSet;
    }

    /**
     * Gets if a pid is available or not based on the value of a bitset
     *
     * @param pid PID to check if available to query
     * @param pidValue parent pid value (hex) which denotes which pids are available
     * @return True if PID is available, false otherwise.
     */
    public static boolean isPidAvailable(final PID pid, final String pidValue) {
        //each parent pid governs 32 pids
        final int pidIndex = Integer.parseInt(pid.PID, 16) % 32 - 1;

        return hexToBitSet(pidValue).get(pidIndex);
    }
}

File: android-obd-library/src/main/java/com/pnuema/android/obd/statics/Translations.kt
package com.pnuema.android.obd.statics

import com.pnuema.android.obd.R
import com.pnuema.android.obd.models.PID
import java.util.*

/**
 * This class provides translations for pids that are based on enumeration or other translation
 * methods other than just conversion using a simple formula.
 *
 * @author Brad Barnhill
 */
object Translations {
    private const val bitsInOneCharValue = 4
    private const val HEX_RADIX = 16

    fun handleSpecialPidEnumerations(pid: PID, buffer: ArrayList<Int>): Boolean {
        if (pid.mode == "01" && pid.PID == "00") {
            mode1Pid00Translation(pid, buffer)
            return true
        } else if (pid.mode == "01" && pid.PID == "01") {
            //dtc count
            mode1Pid01Translation(pid, buffer)
            return true
        } else if (pid.mode == "01" && pid.PID == "51") {
            //fuel type
            mode1Pid51Translation(pid, buffer)
            return true
        }

        return false
    }

    /**
     * Not finished yet
     */
    private fun mode1Pid00Translation(pid: PID, buffer: ArrayList<Int>) {
        if (buffer.size != 6) {
            return
        }

        val hexString = StringBuilder()
        val subList = buffer.subList(2, buffer.size)
        var temp: String
        for (i in subList.indices) {
            temp = Integer.toHexString(subList[i])
            hexString.append(if (temp.length == 1) "0" else "").append(temp)
        }

        pid.calculatedResult = java.lang.Long.parseLong(hexString.toString(), HEX_RADIX).toFloat()
        pid.calculatedResultString = hexString.toString()
    }

    private fun mode1Pid01Translation(pid: PID, buffer: ArrayList<Int>) {
        if (buffer.isNotEmpty() && buffer.size > 2) {
            // ignore first two bytes [hh hh] of the response
            val mil = buffer[2]
            val on = ObdContext.getResourceString(R.string.mode1_pid01_translation_on)
            val off = ObdContext.getResourceString(R.string.mode1_pid01_translation_off)
            val onOff = if (mil and 0x80 == 128) on else off
            pid.calculatedResultString = String.format(ObdContext.getResourceString(R.string.mode1_pid01_translation), onOff, mil and 0x7F)
        }
    }

    private fun mode1Pid51Translation(pid: PID, buffer: ArrayList<Int>) {
        val pidTranslation = ObdContext.getResourceStringArray(R.array.mode1_pid51_translation)
        if (buffer.isNotEmpty() && buffer.size > 2 && buffer[2] < pidTranslation.size) {
            pid.calculatedResultString = pidTranslation[buffer[2]]
        } else {
            pid.calculatedResultString = ObdContext.getResourceString(R.string.pid_value_unavailable)
        }
    }

    /**
     * Convert the given hexadecimal number to bits
     *
     * @param hex The [String] representation of the hexadecimal
     * @return [BitSet] containing the bits of the hexadecimal
     */
    private fun hexToBitSet(hex: String): BitSet {
        val hexBitSize = hex.length * bitsInOneCharValue
        val hexBitSet = BitSet(hexBitSize)

        for (i in hex.indices) {
            val hexChar = hex[i]
            val charBitSet = hexCharToBitSet(hexChar)

            for (j in 0 until bitsInOneCharValue) {
                if (charBitSet.get(j)) {
                    hexBitSet.set(j + (hex.length - i - 1) * bitsInOneCharValue)
                }
            }
        }

        return hexBitSet
    }

    /**
     * Convert the given hexadecimal character to its bits. Note: valid inputs
     * are 0-F
     *
     * @param hexChar the hexadecimal character to convert
     * @return BitSet containing the bits of the hexadecimal character
     */
    private fun hexCharToBitSet(hexChar: Char): BitSet {
        val charBitSet = BitSet(bitsInOneCharValue)
        val hex = Integer.parseInt(hexChar.toString(), HEX_RADIX)

        for (i in 0 until bitsInOneCharValue) {
            val bit = Integer.lowestOneBit(hex shr i)
            if (bit == 1)
                charBitSet.set(i)
        }

        return charBitSet
    }

    /**
     * Gets if a pid is available or not based on the value of a bitset
     *
     * @param pid PID to check if available to query
     * @param pidValue parent pid value (hex) which denotes which pids are available
     * @return True if PID is available, false otherwise.
     */
    fun isPidAvailable(pid: PID, pidValue: String): Boolean {
        //each parent pid governs 32 pids
        val pidIndex = Integer.parseInt(pid.PID, HEX_RADIX) % 32 - 1

        return hexToBitSet(pidValue).get(pidIndex)
    }
}
